commit 292cef5e6262e7976dca310ed9c86f135b807bb4
Author: Dmitry Bogdanov <d.bogdanov@yadro.com>
Date:   Mon Jul 18 18:25:55 2022 +0300

    scsi: target: iscsi: Do not require target authentication
    
    RFC7143 states that Initiator decides what type of authentication to
    use:
    
    The initiator MUST continue with:
        CHAP_N=<N> CHAP_R=<R>
    or, if it requires target authentication, with:
        CHAP_N=<N> CHAP_R=<R> CHAP_I=<I> CHAP_C=<C>
    
    Allow one way authentication if mutual authentication is configured.  That
    passes some tests from Windows HLK for Mutual CHAP with iSNS.
    
    Link: https://lore.kernel.org/r/20220718152555.17084-5-d.bogdanov@yadro.com
    Reviewed-by: Mike Christie <michael.christie@oracle.com>
    Signed-off-by: Dmitry Bogdanov <d.bogdanov@yadro.com>
    Signed-off-by: Martin K. Petersen <martin.petersen@oracle.com>

diff --git a/drivers/target/iscsi/iscsi_target_auth.c b/drivers/target/iscsi/iscsi_target_auth.c
index a5b72968f356..c8a248bd11be 100644
--- a/drivers/target/iscsi/iscsi_target_auth.c
+++ b/drivers/target/iscsi/iscsi_target_auth.c
@@ -416,7 +416,13 @@ static int chap_server_compute_hash(
 	/*
 	 * Get CHAP_I.
 	 */
-	if (extract_param(nr_in_ptr, "CHAP_I", 10, identifier, &type) < 0) {
+	ret = extract_param(nr_in_ptr, "CHAP_I", 10, identifier, &type);
+	if (ret == -ENOENT) {
+		pr_debug("Could not find CHAP_I. Initiator uses One way authentication.\n");
+		auth_ret = 0;
+		goto out;
+	}
+	if (ret < 0) {
 		pr_err("Could not find CHAP_I.\n");
 		goto out;
 	}